src/main.py

```diff
@@ -1,29 +1,93 @@
+    val_dir: str = "../data/LibriSpeech/test-clean"
+    test_dir: str = "../data/LibriSpeech/test-clean"
+
+    # TODO: change this to pre-trained config later
+    model_id: str = "vasudevgupta/tf-wav2vec2-base-960h"
+
+    # wandb args
+    project_name: str = "gsoc-wav2vec2"
+    run_name: str = "finetuning"
+
+
+def main(args):
 
-    model = Wav2Vec2ForCTC.from_pretrained(MODEL_ID)
-    model.freeze_feature_extractor() # for finetuning
-    tr_dataset_fn = partial(DataLoader(data_dir="timit/data/TRAIN"), shuffle=True)
-    val_dataset_fn = partial(DataLoader(data_dir="timit/data/TEST"), shuffle=False)
+    print("######### preparing train set #########")
+    tr_data_args = LibriSpeechDataLoaderArgs(
+        data_dir=args.train_dir,
+        batch_size=args.batch_size,
+        audio_maxlen=args.audio_maxlen,
+        audio_pad_id=0,
+        labels_maxlen=args.labels_maxlen,
+        labels_pad_id=0,
+    )
+    tr_dataset = LibriSpeechDataLoader(tr_data_args)(seed=args.seed)
 
-    T = Trainer(lr=2e-4, max_epochs=2, use_tpu=False)
-    optimizer = tf.keras.optimizers.Adam(lr=T.lr)
-    T.setup(model, optimizer, restore_ckpt=False)
+    print("######### preparing validation set #########")
+    val_data_args = replace(tr_data_args, data_dir=args.val_dir)
+    val_dataset = LibriSpeechDataLoader(val_data_args)(seed=None)
 
-    # let's do training
-    T.train(tr_dataset_fn, val_dataset_fn)
+    print("######### preparing test set #########")
+    test_data_args = replace(val_data_args, data_dir=args.test_dir)
+    test_dataset = LibriSpeechDataLoader(test_data_args)(seed=None)
+
+    print("######### initializing training state #########")
+    input_shape = (args.batch_size, args.audio_maxlen)
+    optimizer = tf.keras.optimizers.Adam(learning_rate=args.lr)
+    loss_tracker = tf.keras.metrics.Mean(name="loss")
+    callbacks = [] # TODO: fill this
+
+    print("######### preparing model #########")
+    model = Wav2Vec2ForCTC.from_pretrained(
+        args.model_id,
+        input_shape=input_shape,
+        apply_spec_augment=args.apply_spec_augment,
+        layer_drop=0, # TODO: layer-drop is not implemented correctly
+    )
+    model.freeze_feature_extractor()  # for finetuning
+    model.summary()
+
+    print("######### initiating training #########")
+    loss_fn = CTCLoss(model.config, input_shape)
+    model.compile(optimizer=optimizer, loss_fn=loss_fn, loss_tracker=loss_tracker)
+
+    # train it like any other TF model; just call `model.fit(...)`
+    model.fit(
+        tr_dataset,
+        validation_data=val_dataset,
+        epochs=args.max_epochs,
+        callbacks=callbacks,
+        verbose="auto",
+    )
+
+    print("######### preparing for evaluation #########")
+    results = model.evaluate(test_dataset, return_dict=True)
+    print("RESULTS:\n", results)
+
+
+if __name__ == "__main__":
+    args = TrainingArgs()
+    tf.random.set_seed(args.seed)
+    main(args)
```

Comment:
For announcements and benchmarking purposes, do we want to try a few different seeds and report the averaged results?

Reply:
Will try different seeds during training stage